File: store.py
"""Private invoice storage; commitments are not claims that an invoice was paid."""
import hashlib,json,os,re,secrets,sqlite3,time,uuid
from pathlib import Path
ROOT=Path(__file__).resolve().parent
DB=Path(os.environ.get('DOCKET_DB',ROOT/'.local/docket.sqlite3'))
class Problem(ValueError):pass

def canonical(value):return json.dumps(value,ensure_ascii=False,sort_keys=True,separators=(',',':'))
def commitment(invoice,salt):return '0x'+hashlib.sha256(('docket:invoice:v1\n'+salt+'\n'+canonical(invoice)).encode()).hexdigest()
def connect():
 DB.parent.mkdir(parents=True,exist_ok=True,mode=0o700)
 c=sqlite3.connect(DB,timeout=15);c.row_factory=sqlite3.Row
 c.execute('CREATE TABLE IF NOT EXISTS records (id TEXT PRIMARY KEY, tx_key TEXT UNIQUE NOT NULL, body TEXT NOT NULL, version INTEGER NOT NULL)')
 os.chmod(DB,0o600);return c

def text(value,name,limit=200):
 if not isinstance(value,str)or not value.strip()or len(value)>limit or any(ord(c)<32 and c not in '\n\t' for c in value):raise Problem(f'{name} is required and must be at most {limit} characters.')
 return value.strip()
def hexvalue(value,name,length):
 if not isinstance(value,str)or not re.fullmatch('0x[0-9a-fA-F]{'+str(length)+'}',value):raise Problem(f'{name} must be a complete hexadecimal value.')
 return value.lower()
def invoice_input(data):
 if not isinstance(data,dict):raise Problem('Send an invoice object.')
 asset=data.get('asset','ETH')
 if asset not in ('ETH','ERC20'):raise Problem('Choose ETH or ERC20.')
 raw=data.get('decimals',18)
 if isinstance(raw,bool)or not str(raw).isdigit()or not 0<=int(raw)<=36:raise Problem('Token decimals must be between 0 and 36.')
 decimals=18 if asset=='ETH' else int(raw)
 amount=text(data.get('amount'),'Expected amount',80)
 if not re.fullmatch(r'[0-9]{1,40}(\.[0-9]{1,36})?',amount):raise Problem('Use a positive decimal amount without commas.')
 whole,_,fraction=amount.partition('.')
 if len(fraction)>decimals:raise Problem('The amount has more decimal places than the token.')
 units=int(whole)*10**decimals+int((fraction or '0').ljust(decimals,'0'))
 if not 0<units<2**256:raise Problem('The amount must be positive and fit the token.')
 return {'reference':text(data.get('reference'),'Invoice reference',100),'counterparty':text(data.get('counterparty'),'Payee name',120),'description':text(data.get('description'),'Work description',2000),'recipient':hexvalue(data.get('recipient'),'Recipient',40),'asset':asset,'token_address':hexvalue(data.get('token_address'),'Token address',40)if asset=='ERC20' else None,'decimals':decimals,'amount':amount,'amount_units':str(units),'source_chain_id':11155111}

def create(data,synthetic=False):
 invoice=invoice_input(data);tx_hash=hexvalue(data.get('tx_hash'),'Transaction hash',64);salt=secrets.token_hex(32)
 record={'id':str(uuid.uuid4()),**invoice,'tx_hash':tx_hash,'invoice':invoice,'salt':salt,'commitment':commitment(invoice,salt),'version':1,'created_at':time.time(),'synthetic':synthetic,'receipt':{'state':'unknown'},'match':{'state':'unknown'},'inclusion':{'state':'unknown'},'registry':{'state':'not_recorded'},'review':{'decision':'unreviewed','note':''},'history':[],'checks':[]}
 record['history'].append({'at':record['created_at'],'action':'Record created','version':1})
 try:
  with connect()as c:c.execute('INSERT INTO records VALUES (?,?,?,?)',(record['id'],str(synthetic)+':11155111:'+tx_hash,canonical(record),1))
 except sqlite3.IntegrityError:raise Problem('This transaction is already linked to a record. Review that record instead.')
 return public_view(record)

def get(id):
 with connect()as c:row=c.execute('SELECT body FROM records WHERE id=?',(id,)).fetchone()
 if not row:raise Problem('Record not found. Refresh the list.')
 return json.loads(row['body'])
def all_records():
 with connect()as c:rows=c.execute('SELECT body FROM records ORDER BY rowid DESC').fetchall()
 return [public_view(json.loads(row['body']))for row in rows]
def public_view(record):
 return {k:v for k,v in record.items()if k not in ('salt','invoice')}
def update(id,version,values,action):
 with connect()as c:
  c.execute('BEGIN IMMEDIATE');row=c.execute('SELECT body,version FROM records WHERE id=?',(id,)).fetchone()
  if not row or row['version']!=version:raise Problem('This record changed. Refresh and review it before trying again.')
  r=json.loads(row['body']);
  if 'receipt'in values:
   r.setdefault('checks',[]).append({'at':time.time(),'receipt':values['receipt'],'match':values['match'],'inclusion':values['inclusion']});r['checks']=r['checks'][-20:]
  r.update(values);r['version']+=1;r['history'].append({'at':time.time(),'action':action,'version':r['version']})
  c.execute('UPDATE records SET body=?,version=? WHERE id=?',(canonical(r),r['version'],id))
 return public_view(r)
def review(id,version,data):
 decision=data.get('decision')
 if decision not in ('needs_work','checked'):raise Problem('Choose a review decision.')
 r=get(id);note=text(data.get('note'),'Review note',2000)
 if decision=='checked' and not(r['receipt']['state']=='success'and r['match']['state']=='matched'and r['inclusion']['state']=='verified'and not r['synthetic']):raise Problem('Check the source receipt, payment match and real inclusion proof before marking reviewed.')
 return update(id,version,{'review':{'decision':decision,'note':note}},'Review saved')
def export_record(id,private=False):
 r=get(id)
 base={k:r[k]for k in ('id','commitment','tx_hash','source_chain_id','synthetic','receipt','match','inclusion','registry','created_at')}
 base['checks']=r.get('checks',[])
 # Notes and invoice text are deliberately absent from the default public export.
 base.update(format='docket-evidence-v1',privacy='public',claim='Inclusion and RPC payment checks are separate. No invoice ownership or payment guarantee.')
 if private:base.update(privacy='private',invoice=r['invoice'],salt=r['salt'],review=r['review'],history=r['history'])
 return base

def sample():
 return create({'reference':'SAMPLE-014','counterparty':'Campus design team (sample)','description':'Synthetic example: a failed source call must never look like a paid invoice.','recipient':'0x'+'2'*40,'amount':'0.025','asset':'ETH','tx_hash':'0x'+'a'*64},True)

File: public-app/store.mjs
import {sha256,toUtf8Bytes,randomBytes,hexlify} from 'ethers';
export const KEY='docket:browser:v1';
export function canonical(value){if(Array.isArray(value))return '['+value.map(canonical).join(',')+']';if(value&&typeof value==='object')return '{'+Object.keys(value).sort().map(k=>JSON.stringify(k)+':'+canonical(value[k])).join(',')+'}';return JSON.stringify(value);}
export const commitment=(invoice,salt)=>sha256(toUtf8Bytes('docket:invoice:v1\n'+salt+'\n'+canonical(invoice)));
function text(value,name,max){if(typeof value!=='string'||!value.trim()||value.length>max||/[\x00-\x08\x0b\x0c\x0e-\x1f]/.test(value))throw Error(`${name} is required and must be at most ${max} characters.`);return value.trim();}
function hex(value,name,n){if(typeof value!=='string'||!new RegExp('^0x[0-9a-fA-F]{'+n+'}$').test(value))throw Error(`${name} must be a complete hexadecimal value.`);return value.toLowerCase();}
export function invoiceInput(data){
 if(!data||typeof data!=='object'||Array.isArray(data))throw Error('Send an invoice object.');
 const asset=data.asset||'ETH';if(!['ETH','ERC20'].includes(asset))throw Error('Choose ETH or ERC20.');
 if(!/^\d+$/.test(String(data.decimals??18))||Number(data.decimals??18)>36)throw Error('Token decimals must be between 0 and 36.');
 const decimals=asset==='ETH'?18:Number(data.decimals??18),amount=text(data.amount,'Expected amount',80);
 if(!/^[0-9]{1,40}(\.[0-9]{1,36})?$/.test(amount))throw Error('Use a positive decimal amount without commas.');
 const [whole,fraction='']=amount.split('.');if(fraction.length>decimals)throw Error('The amount has more decimal places than the token.');
 const units=BigInt(whole)*10n**BigInt(decimals)+BigInt(fraction.padEnd(decimals,'0')||'0');if(units<=0n||units>=2n**256n)throw Error('The amount must be positive and fit the token.');
 return {reference:text(data.reference,'Invoice reference',100),counterparty:text(data.counterparty,'Payee name',120),description:text(data.description,'Work description',2000),recipient:hex(data.recipient,'Recipient',40),asset,token_address:asset==='ERC20'?hex(data.token_address,'Token address',40):null,decimals,amount,amount_units:String(units),source_chain_id:11155111};
}
export class BrowserStore{
 constructor(storage){this.storage=storage;}
 read(){const data=this.storage.getItem(KEY);if(!data)return [];const rows=JSON.parse(data);if(!Array.isArray(rows))throw Error('The browser record store is invalid. Export a backup before repairing it.');return rows;}
 save(rows){this.storage.setItem(KEY,JSON.stringify(rows));}
 get(id){const r=this.read().find(r=>r.id===id);if(!r)throw Error('Record not found. Refresh the list.');return r;}
 all(){return this.read().slice().reverse().map(r=>this.view(r));}
 view(r){const {invoice,salt,...visible}=r;return visible;}
 create(data,synthetic=false,restored=null){const invoice=invoiceInput(data),tx_hash=hex(data.tx_hash,'Transaction hash',64),rows=this.read();if(rows.some(r=>r.synthetic===synthetic&&r.tx_hash===tx_hash))throw Error('This transaction is already linked to a record. Review it instead.');const salt=hexlify(randomBytes(32)).slice(2),at=Date.now()/1000,r={id:crypto.randomUUID(),...invoice,tx_hash,invoice,salt,commitment:commitment(invoice,salt),version:1,created_at:at,synthetic,receipt:{state:'unknown'},match:{state:'unknown'},inclusion:{state:'unknown'},registry:{state:'not_recorded'},review:{decision:'unreviewed',note:''},history:[{at,action:'Record created',version:1}],checks:[]};if(restored){Object.assign(r,restored);r.history.push({at,action:'Private backup restored. Run fresh checks before reviewing.',version:1});}rows.push(r);this.save(rows);return this.view(r);}
 update(id,version,values,action){const rows=this.read(),r=rows.find(r=>r.id===id);if(!r||r.version!==version)throw Error('This record changed. Refresh and review it before trying again.');if(values.receipt){r.checks.push({at:Date.now()/1000,receipt:values.receipt,match:values.match,inclusion:values.inclusion});r.checks=r.checks.slice(-20);}Object.assign(r,values);r.version++;r.history.push({at:Date.now()/1000,action,version:r.version});this.save(rows);return this.view(r);}
 review(id,version,data){const r=this.get(id),decision=data.decision,note=text(data.note,'Review note',2000);if(!['needs_work','checked'].includes(decision))throw Error('Choose a review decision.');if(decision==='checked'&&(r.synthetic||r.receipt.state!=='success'||r.match.state!=='matched'||r.inclusion.state!=='verified'))throw Error('Check the receipt, payment match and real inclusion proof before marking reviewed.');return this.update(id,version,{review:{decision,note}},'Review saved');}
 restore(bundle){
  if(!bundle||bundle.format!=='docket-evidence-v1'||bundle.privacy!=='private')throw Error('Choose a Docket private backup. Public evidence has no invoice details.');
  const invoice=invoiceInput(bundle.invoice),salt=String(bundle.salt||'');if(!/^[a-f0-9]{64}$/.test(salt)||commitment(invoice,salt)!==bundle.commitment)throw Error('The private backup does not match its invoice commitment.');
  const note=bundle.review?.note?text(bundle.review.note,'Review note',2000):'';
  const saved={review:{decision:'unreviewed',note},salt,commitment:bundle.commitment};
  
  const prior=bundle.registry;
  if(prior&&['pending','recorded','recovery'].includes(prior.state)){
   // These are location hints only. Fresh source proof and registry RPC checks are required.
   const hint={state:'recovery',hash:prior.hash?hex(prior.hash,'Registry transaction hash',64):null};
   if(prior.from||prior.submitter)hint.from=hex(prior.from||prior.submitter,'Registry submitter',40);
   for(const key of ['nonce','startBlock'])if(Number.isSafeInteger(prior[key])&&prior[key]>=0)hint[key]=prior[key];
   saved.registry=hint;
  }
  return this.create({...invoice,tx_hash:bundle.tx_hash},bundle.synthetic===true,saved);
 }
 export(id,privateBundle=false){const r=this.get(id),out={};for(const k of ['id','commitment','tx_hash','source_chain_id','synthetic','receipt','match','inclusion','registry','created_at','checks'])out[k]=r[k];Object.assign(out,{format:'docket-evidence-v1',privacy:'public',claim:'Inclusion and RPC payment checks are separate. No invoice ownership or payment guarantee.'});if(privateBundle)Object.assign(out,{privacy:'private',invoice:r.invoice,salt:r.salt,review:r.review,history:r.history});return out;}
}
